ksim/observation.py

```diff
@@ -230,25 +231,35 @@ def create(
     def get_name(self) -> str:
         return f"{self.sensor_name}_obs"
 
-    def observe(self, state: PhysicsData, rng: PRNGKeyArray) -> Array:
-        sensor_data = state.sensordata[self.sensor_idx_range[0] : self.sensor_idx_range[1]].ravel()
+    def observe(self, rollout_state: RolloutVariables, rng: PRNGKeyArray) -> Array:
+        sensor_data = rollout_state.physics_state.data.sensordata[
+            self.sensor_idx_range[0] : self.sensor_idx_range[1]
+        ].ravel()
         return sensor_data
 
     def add_noise(self, observation: Array, rng: PRNGKeyArray) -> Array:
         return add_noise(observation, rng, self.noise_type, self.noise)
 
 
+@attrs.define(frozen=True)
+class HistoryObservation(Observation):
+    def observe(self, rollout_state: RolloutVariables, rng: PRNGKeyArray) -> Array:
+        if not isinstance(rollout_state.carry, Array):
+            raise ValueError(f"History observation carry must be an array, got {type(rollout_state.carry)}")
+        return rollout_state.carry
```

Comment:
hmm - okay, so i feel like rather than having this here and type checking the carry, we should just let downstream users implement whatever function themselves. this feels hacky otherwise